quodlibet/quodlibet/ext/playorder/shufflebygrouping.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Phoenix Dailey
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+import random
+
+from gi.repository import Gtk, GLib
+
+from quodlibet import _
+from quodlibet import app
+from quodlibet.order import OrderInOrder
+from quodlibet.order import OrderRemembered
+from quodlibet.order.reorder import Reorder
+from quodlibet.plugins import PluginConfig
+from quodlibet.plugins.playorder import ShufflePlugin
+from quodlibet.qltk import Icons
+
+
+pconfig = PluginConfig("shufflebygrouping")
+pconfig.defaults.set("grouping", "~grouping~album~albumartist")
+pconfig.defaults.set("grouping_test", "grouping")
+pconfig.defaults.set("delay", 0)
+
+
+class ShuffleByGrouping(ShufflePlugin, OrderInOrder, OrderRemembered):
+    PLUGIN_ID = "shufflebygrouping"
+    PLUGIN_NAME = _("Shuffle by Grouping")
+    PLUGIN_DESC = _("Shuffles by a grouping of songs defined by a common tag "
+                    "instead of by track, similar to album shuffle. This is "
+                    "useful for shuffling multi-movement classical works "
+                    "while making sure all movements play in order before "
+                    "shuffling to the next piece.")
+    PLUGIN_ICON = Icons.MEDIA_PLAYLIST_SHUFFLE
+    display_name = _("Shuffle by grouping")
+    priority = Reorder.priority
+
+    def next(self, playlist, current):
+        return self._next(playlist, current)
+
+    def _next(self, playlist, current, delay_on=True):
+        grouping = str(pconfig.gettext("grouping")).strip()
+        grouping_test = str(pconfig.gettext("grouping_test")).strip()
+        delay = pconfig.getint("delay")
+
+        # Keep track of played songs
+        OrderRemembered.next(self, playlist, current)
+        remaining = OrderRemembered.remaining(self, playlist)
+
+        # Check if playlist is finished or empty
+        if len(remaining) <= 0:
+            OrderRemembered.reset(self, playlist)
+            return None
+
+        # Play next song in current grouping
+        next_song = super().next(playlist, current)
+        if (current is not None and next_song is not None and
+            self._tag_defined(grouping_test, playlist, next_song) and
+            self._same_tag(grouping, playlist, current, next_song)):
+            return next_song
+
+        # Pause for a moment before picking new group
+        if delay_on and delay > 0:
+            app.player.paused = True
```

Comment:
BTW check the Random Album plugin for a nice UI-linked way of doing this (which is cancellable by the user)

Reply:
What is the purpose of this try-except block?
Lines 20-23 of [randomalbum.py](https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/ext/events/randomalbum.py#L20):
```
try:
    from quodlibet.qltk import notif, Icons
except Exception:
    notif = None
```